### PR TITLE
feat(fastify): Support `{hookName: 'onRequest'}`

### LINKED ISSUE #1350: `@clerk/fastify` should hook earlier in request lifecycle
- [x] Review the documentation: https://clerk.com/docs
- [x] Search for existing issues: https://github.com/clerkinc/javascript/issues
- [x] Go through package changelog files.
- [x] Provide the Frontend API key from your application dashboard.

## Package + Version

- [x] `@clerk/fastify`

## Description

Currently, `@clerk/fastify` adds a lifecycle hook to `preHandler`, the hook immediately preceding the API handler code. However, this means that fastify runs both `parsing` and `validation` steps to the request before Clerk confirms whether or not the request is even valid.

I think the plugin should hook either `onRequest` or `preParsing` ([docs](https://www.fastify.io/docs/latest/Reference/Hooks/)). By default, this means the plugin will provide maximum performance. It'll also allow user plugins to select how early in the lifecycle they implement authz. 

If this is too much of a breaking change, then a good compromise is to offer configuration as to which hook to leverage.

Thoughts?

### PR DESCRIPTION
**Pull request by mikestopcontinues**
## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [x] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

Adds support for option `{hookName: 'onRequest'}`, which binds the callback to the specified hook.

By default, or when specified as such, the callback will continue to be bound to `preHandler`.

This update prevents binding to any other hook to minimize compatibility requirements going forward.

Fixes #1350

### CODE REVIEW

File: packages/fastify/src/clerkPlugin.ts [hunk 1]
```diff
@@ -7,7 +7,19 @@ import { withClerkMiddleware } from './withClerkMiddleware';
 const plugin: FastifyPluginCallback = (instance: FastifyInstance, opts: ClerkFastifyOptions, done) => {
   instance.decorateRequest('auth', null);
   // run clerk as a middleware to all scoped routes
-  instance.addHook('preHandler', withClerkMiddleware(opts));
+  switch (opts.hookName) {
```
[the remark this answers is not in the document]
Comment (dimkl) [reply packages/fastify/src/clerkPlugin.ts:10]: ❓ Would this also cause a type error ?
```typescript

const ALLOWED_HOOKS = ['onRequest', 'preHandler'] as const;

const plugin: FastifyPluginCallback = (instance: FastifyInstance, opts: ClerkFastifyOptions, done) => {
// ...
const hookname = opts.hookName || 'preHandler';
if (!ALLOWED_HOOKS.includes(hookName)) {
    throw new Error(`Unsupported hookName: ${hookName}`);
}

instance.addHook(hookName, withClerkMiddleware(opts));
// ...
}

File: package-lock.json [hunk 2]
```diff
@@ -41321,10 +41321,10 @@
     },
     "packages/backend": {
       "name": "@clerk/backend",
-      "version": "0.21.0",
```
Comment (dimkl) [inline package-lock.json:41324]: 🔧 Please revert the changes in the `package-lock.json` file. We will make the appropriate change.

**Review by dimkl — COMMENTED** [review]
Nice work!! 🎉 
I believe that we are ready to merge after addressing my minor comments.
Thank you for your contribution @mikestopcontinues! 😄

File: packages/fastify/src/clerkPlugin.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikestopcontinues) [reply packages/fastify/src/clerkPlugin.ts:10]: Yes, this is exactly what causes the error. The type of `opts.hookName` is already `'onRequest' | 'preHandler'`. I've encountered this issue in TS before. Discrimination doesn't work as expected with overloaded functions. So you either have to split up the union or use `@ts-expect-error`.

Here's the error:
```
No overload matches this call.
  The last overload gave the following error.
    Argument of type '"onRequest" | "preHandler"' is not assignable to parameter of type '"onClose"'.
      Type '"onRequest"' is not assignable to type '"onClose"'.ts(2769)
instance.d.ts(467, 3): The last overload is declared here.
```

And here's the type that causes the issue:
```
type Hooks = Parameters<FastifyInstance['addHook']>[0]; // === 'onClose' SERIOUSLY!
```

Anyway, I updated the code to read more simply. I hope that works!

**Comment by mikestopcontinues** [comment]
That's exactly what causes the error. The type of `opts.hookName` is already `'onRequest' | 'preHandler'`. I've encountered this issue in TS before. Discrimination doesn't work as expected with overloaded functions. So you either have to split up the union or use `@ts-expect-error`.

Here's the error:
```
No overload matches this call.
  The last overload gave the following error.
    Argument of type '"onRequest" | "preHandler"' is not assignable to parameter of type '"onClose"'.
      Type '"onRequest"' is not assignable to type '"onClose"'.ts(2769)
instance.d.ts(467, 3): The last overload is declared here.
```

And here's the type that causes the issue:
```
type Hooks = Parameters<FastifyInstance['addHook']>[0]; // === 'onClose' SERIOUSLY!
```

Anyway, I updated the code to read more simply. I hope that works!

**Review by nikosdouvlis — COMMENTED** [review]
A minor comment from me, this is good to go.
Many thanks @mikestopcontinues :)

**Review by nikosdouvlis — APPROVED** [review]

**Review by dimkl — APPROVED** [review]

**Comment by nikosdouvlis** [comment]
Merging so we can get this to staging. 
Thank you @mikestopcontinues

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
